src/Swarm/TUI/Model/Achievement/Persistence.hs

```diff
@@ -4,38 +4,73 @@ module Swarm.TUI.Model.Achievement.Persistence where
 
 import Control.Arrow (left)
 import Control.Carrier.Lift (sendIO)
+import Control.Monad (forM, forM_)
+import Data.Either (partitionEithers)
 import Data.Text (Text, pack)
+import Data.Text qualified as T
 import Data.Yaml qualified as Y
 import Data.Yaml.Aeson (prettyPrintParseException)
 import Swarm.TUI.Model.Achievement.Attainment
 import Swarm.Util
 import System.Directory (
+  createDirectoryIfMissing,
+  doesDirectoryExist,
   doesFileExist,
+  listDirectory,
  )
 import System.FilePath
+import System.IO (hPutStrLn, stderr)
 
 -- | Get path to swarm achievements, optionally creating necessary
 --   directories.
 getSwarmAchievementsPath :: Bool -> IO FilePath
-getSwarmAchievementsPath createDirs =
-  (</> "achievements") <$> getSwarmDataPath createDirs
+getSwarmAchievementsPath createDirs = do
+  dp <- getSwarmDataPath False
+  let achievementsDir = dp </> "achievement"
+  createDirectoryIfMissing createDirs achievementsDir
+  return achievementsDir
 
 -- | Load saved info about achievements from XDG data directory.
 loadAchievementsInfo ::
   IO (Either Text [Attainment])
 loadAchievementsInfo = do
-  savedAchievementsPath <- getSwarmAchievementsPath True
-  hasSavedAchievements <- sendIO $ doesFileExist savedAchievementsPath
-  if not hasSavedAchievements
-    then return $ pure []
-    else
-      left (pack . prettyPrintParseException)
-        <$> sendIO (Y.decodeFileEither savedAchievementsPath)
+  savedAchievementsPath <- getSwarmAchievementsPath False
+  doesParentExist <- doesDirectoryExist savedAchievementsPath
+  if doesParentExist
+    then do
+      contents <- listDirectory savedAchievementsPath
+      eithersList <- forM contents $ \p -> do
+        let fullPath = savedAchievementsPath </> p
+        isFile <- doesFileExist fullPath
+        if isFile
+          then do
+            eitherDecodedFile <- sendIO (Y.decodeFileEither fullPath)
+            return $ left (pack . prettyPrintParseException) eitherDecodedFile
+          else
+            return $
+              Left $
+                T.unwords
+                  [ "Entry"
+                  , quote $ T.pack fullPath
+                  , "is not a file!"
+                  ]
+      let (bads, goods) = partitionEithers eithersList
+      forM_ bads $ \b ->
+        hPutStrLn stderr $ T.unpack b
+      return $ pure goods
+    else return $ pure []
 
 -- | Save info about achievements to XDG data directory.
 saveAchievementsInfo ::
   [Attainment] ->
   IO ()
-saveAchievementsInfo si = do
+saveAchievementsInfo attainmentList = do
   savedAchievementsPath <- getSwarmAchievementsPath True
-  Y.encodeFile savedAchievementsPath si
+  forM_ attainmentList $ \x -> do
+    let achievementName =
+          T.unpack $
+            T.takeWhileEnd (/= ' ') $
+              T.pack $
+                show (_achievement x)
```

Comment:
This seems like kind of an ugly and nonobvious way to extract a name for an achievement.  I had to go look at the data type definitions to figure out what this is even doing.  Why not just pattern-match on the `CategorizedAchievement` and call `show` on the contents?